File: leetcode/python/0066_plus_one.py
class Solution(object):
    def plusOne(self, digits):
        """
        :type digits: List[int]
        :rtype: List[int]
        """
        # if digits[-1] + 1 < 10: digits[-1] += 1, return digits
        # else:
        # carry = 1, digits[-1] = (digits[-1] + 1) % 10
        # loop through digits from -2 position to 0 using i:
        # -> if digits[i] + carry < 10: digits[i] += carry, carry = 0, break
        # -> elif digits[i] + carry > 10: digits[i] = (digits[i] + carry) % 10
        # if carry == 1: digits.insert(1)
        # return digits

        if digits[-1] + 1 < 10:
            digits[-1] += 1
            return digits

        carry = 1
        digits[-1] = (digits[-1] + 1) % 10

        for i in range(len(digits) - 2, -1, -1):
            if digits[i] + carry < 10:
                digits[i] = (digits[i] + carry) % 10
                carry = 0
                break
            elif digits[i] + carry >= 10:
                digits[i] = (digits[i] + carry) % 10
            
        if carry == 1:
            digits.insert(0, 1)
        return digits

File: leetcode/java/0066_plus_one.java
class Solution {
    public int[] plusOne(int[] digits) {
        int carry = 1;
    
        if (digits[digits.length - 1] < 9) {
            digits[digits.length - 1] += 1;
            return digits;
        }

        for (int i = digits.length - 1; i >= 0; i--) {
            if (digits[i] == 9) digits[i] = 0;
            else if (digits[i] < 9 && carry == 1) {
                digits[i] += 1;
                carry = 0;
            }

            if (carry == 0) return digits;
            if (i == 0) {
                digits = new int[digits.length + 1];
                digits[0] = 1;
            }
        }

        return digits;
    }
}